File: LAB_10/State.cpp
#include <iostream>
#include <string>

class State1; 

class ATM {
    private:
        std::string state;

    public:
        void setState(const std::string& newState) {
            state = newState;
        }

        std::string getState() const {
            return state;
        }

        void changeState(State1* state);
};

class State1 {
    public:
        virtual void handle(ATM& atm) = 0;
};

void ATM::changeState(State1* state) {
    state->handle(*this);
}


class WaitingForCardState : public State1 {
public:
    void handle(ATM& atm) override {
        atm.setState("Waiting for card");
    }
};

class PINVerificationState : public State1 {
public:
    void handle(ATM& atm) override {
        atm.setState("Checking PIN");
    }
};

class WithdrawalState : public State1 {
public:
    void handle(ATM& atm) override {
        atm.setState("Cash withdrawal");
    }
};

int main() {
    ATM* atm = new ATM();
    WaitingForCardState* waitingForCard = new WaitingForCardState();
    PINVerificationState* verifyPIN = new PINVerificationState();
    WithdrawalState* withdrawal = new WithdrawalState();

    atm->changeState(waitingForCard);
    std::cout << atm->getState() << std::endl;

    atm->changeState(verifyPIN);
    std::cout << atm->getState() << std::endl;

    atm->changeState(withdrawal);
    std::cout << atm->getState() << std::endl;

    delete atm;
    delete waitingForCard;
    delete verifyPIN;
    delete withdrawal;

    return 0;
}

File: LAB_10/State.java
class ATM {
    private String state;

    public void setState(String state) {
        this.state = state;
    }

    public String getState() {
        return state;
    }

    public void changeState(State1 state) {
        state.handle(this);
    }
}

interface State1 {
    void handle(ATM atm);
}

class WaitingForCardState implements State1 {
    @Override
    public void handle(ATM atm) {
        atm.setState("Waiting for card");
    }
}

class PINVerificationState implements State1 {
    @Override
    public void handle(ATM atm) {
        atm.setState("Checking PIN");
    }
}

class WithdrawalState implements State1 {
    @Override
    public void handle(ATM atm) {
        atm.setState("Cash withdrawal");
    }
}



public class State {
    public static void main(String[] args) {
        ATM atm = new ATM();

        State1 waitingForCard = new WaitingForCardState();
        State1 verifyPIN = new PINVerificationState();
        State1 withdrawal = new WithdrawalState();

        atm.changeState(waitingForCard);
        System.out.println(atm.getState());

        atm.changeState(verifyPIN);
        System.out.println(atm.getState());

        atm.changeState(withdrawal);
        System.out.println(atm.getState());
    }
}